src/lib.rs

```diff
@@ -282,12 +282,33 @@ pub extern "C" fn setsockopt(
     optval: *const c_void,
     optlen: socklen_t,
 ) -> c_int {
-}
+}*/
 
 #[no_mangle]
-pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {}
+pub extern "C" fn listen(fd: c_int, n: c_int) -> c_int {
+    let lock = match INDEX.get(fd) {
+        Some(s) => s,
+        None => return next!(listen(fd, n)),
+    };
 
-#[no_mangle]
+    let mut sock = lock.write().unwrap();
+
+    match *sock {
+        Socket::Bound(p) => {
+            if next!(listen(fd, n)) == -1 {
+                return -1;
+            }
+
+            *sock = Socket::Listening(p);
+        },
+
+        _ => return error(libc::EBADFD, -1),
+    }
+
+    0
```

Comment:
Move the `0` into line 302. Then you can remove the `return` keyword on line 305.